google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java

```diff
@@ -476,6 +477,29 @@ static String getServiceAccountProjectId(String credentialsPath) {
     return project;
   }
 
+  /* 
+   * Returns true if the projectId is valid. This method checks whether the projectId
+   * contains only lowercase letters, digits and hyphens, starts with a lowercase letter
+   * and does not end with a hyphen, but does not check the length of projectId. This
+   * method is primarily used to protect against DNS hijacking.
+   */
+  static boolean isValidProjectId(String projectId) {   
+    for (char c : projectId.toCharArray()) {
```

Comment:
This doesn't validate that the *first* character isn't a hyphen, either. Also, some project IDs have colons in them (i.e. Googleplex apps). This validation will reject them, incorrectly. What are the constraints here based on? Is there a spec somewhere?

Reply:
There are constraints about project-id here[https://github.com/GoogleCloudPlatform/google-cloud-java/blob/536af840f93622a8743d7c129104e72909aeab0a/google-cloud-resourcemanager/src/main/java/com/google/cloud/resourcemanager/ProjectInfo.java#L18] and here[https://cloud.google.com/resource-manager/reference/rest/v1beta1/projects]. Also trying to create a project through cloud console with a project-id that does not follow the constraints will not be allowed. I haven't checked Googleplex apps though. I'll make the changes if colons are allowed in that case.

Reply:
Neither of those is an actual, documented constraint. They're both implementation details of APIs. How do you know that pre-existing projects meet those constraints, for example? Assuming _isn't_ enough, especially when it will break customers, as *this change will* for Googleplex apps.